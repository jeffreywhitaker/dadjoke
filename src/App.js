// import dependencies
import React, { useEffect } from 'react'
import { connect } from 'react-redux'
import { Route, Redirect } from 'react-router-dom'
import styled from 'styled-components'

// import components
import NavBar from './components/NavBar'
import PublicJokes from './components/PublicJokes'
import PrivateJokes from './components/PrivateJokes'
import Login from './components/Login'
import Signup from './components/Signup'
import AddJoke from './components/AddJoke'
import BottomNav from './components/BottomNav'

import { checkTokenValidity } from './actions/actions'

// App component
export function App(props) {
  // destructure props
  const { checkTokenValidity } = props

  // use effect to check for token
  useEffect(() => {
    if (localStorage.token) {
      checkTokenValidity()
    }
  }, [checkTokenValidity])

  // return components
  return (
    <AppWrapper className="App">
      <Route component={NavBar} />
      <ContentContainer>
        <Redirect exact path="/" to="/publicjokes" />
        <Route path="/publicjokes" component={PublicJokes} />
        <Route path="/privatejokes" component={PrivateJokes} />
        <Route path="/login" component={Login} />
        <Route path="/signup" component={Signup} />
        <Route path="/publicjokes/add" component={AddJoke} />
        <Route path="/privatejokes/add" component={AddJoke} />
      </ContentContainer>
      <Route component={BottomNav} />
    </AppWrapper>
  )
}

// export component
export default connect(null, { checkTokenValidity })(App)

// styled components
const AppWrapper = styled.div`
  max-width: 1100px;
  width: 100%;
  height: 100%;
  min-height: 100vh;
  margin: 0 auto;
  background: lightslategray;
  display: flex;
  flex-direction: column;
`

// set max width to 1100px eventually

const ContentContainer = styled.article`
  flex-grow: 1;
`
