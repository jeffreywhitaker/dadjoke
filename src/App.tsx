// import dependencies
import React, { useEffect, useState } from 'react'
import { connect, ConnectedProps } from 'react-redux'
import { Route, Redirect } from 'react-router-dom'
import styled from 'styled-components'

// import components
import Header from './components/Header'
import IntroModal from './components/modals/IntroModal'
import JokesWrapper from './components/JokesWrapper'
import Login from './components/Login'
import MessageBoard from './components/messageBoard'
import Profile from './components/Profile/Profile'
import Signup from './components/Signup'
import ThreadView from './components/MessageBoard/ThreadView'

import { ifSessionExistsLogIn, userLogin } from './actions/actions'

// App component
function App(props: Props) {
  // destructure props
  const { ifSessionExistsLogIn, isLoggedIn, userLogin } = props
  const [isDarkModeOn, setIsDarkModeOn] = useState(false)
  // for intro modal
  const [showModal, setShowModal] = useState(false)
  const handleClose = () => {
    setShowModal(false)
  }

  function handleDemo() {
    console.log('handleDemo')
    userLogin({ username: 'testuser', password: 'password' })
    setShowModal(false)
  }

  function handleDoNotShowAgain() {
    localStorage.setItem('doNotShowIntroModal', 'true')
    setShowModal(false)
  }

  // use effect to check for token
  useEffect(() => {
    if (localStorage.getItem('dark-mode') === 'true') {
      setIsDarkModeOn(true)
    }

    ifSessionExistsLogIn()
    if (localStorage.getItem('doNotShowIntroModal') !== 'true') {
      setShowModal(true)
    }
  }, [ifSessionExistsLogIn])

  const toggleDarkMode = () => {
    console.log('dark mode is: ', isDarkModeOn)
    localStorage.setItem('dark-mode', isDarkModeOn.toString())
    setIsDarkModeOn(!isDarkModeOn)
  }

  // return components
  return (
    <Main isDarkModeOn={isDarkModeOn}>
      <Header isDarkModeOn={isDarkModeOn} toggleDarkMode={toggleDarkMode} />
      <AppWrapper className="App">
        <Route exact path="/">
          <Redirect to="/publicjokes" />
        </Route>
        <Route path="/publicjokes" component={JokesWrapper} />
        <Route path="/privatejokes" component={JokesWrapper} />
        <Route path="/mboard/:threadId" component={ThreadView} />
        <Route path="/mboard" component={MessageBoard} exact />
        <Route path="/login" component={Login} />
        <Route path="/signup" component={Signup} />
        <Route path="/profile/:username" component={Profile} />
      </AppWrapper>

      <IntroModal
        handleClose={handleClose}
        handleDemo={handleDemo}
        handleDoNotShowAgain={handleDoNotShowAgain}
        isLoggedIn={isLoggedIn}
        setShowModal={setShowModal}
        showModal={showModal}
      />
    </Main>
  )
}

interface LoginReducer {
  isLoggedIn: boolean
}

// connect component to redux store
const mapStateToProps = (state: { loginReducer: LoginReducer }) => {
  return {
    isLoggedIn: state.loginReducer.isLoggedIn,
  }
}

// export component
const connector = connect(mapStateToProps, { ifSessionExistsLogIn, userLogin })
type PropsFromRedux = ConnectedProps<typeof connector>
type Props = PropsFromRedux
export default connector(App)

// styled components
interface MainProps {
  isDarkModeOn: boolean
}
const Main = styled.main`
  background-color: ${(props: MainProps) =>
    props.isDarkModeOn ? 'darkgray' : 'white'};
`

const AppWrapper = styled.section`
  max-width: 1100px;
  width: 100%;
  height: 100%;
  min-height: 100vh;
  margin: 0 auto;
  /* background: lightslategray; */
`
