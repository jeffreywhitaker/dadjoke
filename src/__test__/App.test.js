// import React from 'react'
// import ReactDOM from 'react-dom'
import { BrowserRouter as Router } from 'react-router-dom'
// import { App } from '../App'

// it('renders without crashing', () => {
//   const div = document.createElement('div')
//   ReactDOM.render(<Router><App props={() => {}} /></Router>, div)
//   ReactDOM.unmountComponentAtNode(div)
// })

import React from 'react'
import { App } from '../App'
import ShallowRenderer from 'react-test-renderer/shallow'
// import 'jest-styled-components';

describe('App component', () => {
  it('renders correctly', () => {
    // jest.mock('react-router-dom', () => ({
    //   useHistory: () => ({
    //     push: jest.fn(),
    //   }),
    // }))

    const renderer = new ShallowRenderer()
    renderer.render(<App props={() => {}} />)

    expect(renderer.getRenderOutput()).toMatchSnapshot()
  })
})
