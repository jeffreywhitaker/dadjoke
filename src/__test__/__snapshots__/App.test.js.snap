// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders correctly 1`] = `
<ForwardRef(styled.div)
  className="App"
>
  <Route
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(NavBar)",
        "type": [Function],
      }
    }
  />
  <ForwardRef(styled.article)>
    <Redirect
      exact={true}
      path="/"
      to="/login"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(PublicJokes)",
          "type": [Function],
        }
      }
      path="/publicjokes"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(PrivateJokes)",
          "type": [Function],
        }
      }
      path="/privatejokes"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(Login)",
          "type": [Function],
        }
      }
      path="/login"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(Signup)",
          "type": [Function],
        }
      }
      path="/signup"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(AddJoke)",
          "type": [Function],
        }
      }
      path="/publicjokes/add"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(AddJoke)",
          "type": [Function],
        }
      }
      path="/privatejokes/add"
    />
  </ForwardRef(styled.article)>
  <Route
    component={[Function]}
  />
</ForwardRef(styled.div)>
`;
